device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java

```diff
@@ -39,51 +39,208 @@
     private static final String SSL_PREFIX = "ssl://";
     private static final String SSL_PORT_SUFFIX = ":8883";
 
-    private static final String ModelIdParam = "model-id";
+    private static final int KEEP_ALIVE_INTERVAL = 230;
+    private static final int MQTT_VERSION = 4;
+    private static final boolean SET_CLEAN_SESSION = false;
+
+    private static final String MODEL_ID = "model-id";
 
     private String connectionId;
     private String webSocketQueryString;
-
+    private final Object mqttConnectionStateLock = new Object(); // lock for preventing simultaneous open and close calls
+    private final DeviceClientConfig config;
+    private IotHubConnectionStatus state = IotHubConnectionStatus.DISCONNECTED;
     private IotHubListener listener;
 
-    //Messaging clients
-    private MqttMessaging deviceMessaging;
-    private MqttDeviceTwin deviceTwin;
-    private MqttDeviceMethod deviceMethod;
+    //Messaging clients, never null
+    private final MqttMessaging deviceMessaging;
+    private final MqttDeviceTwin deviceTwin;
+    private final MqttDeviceMethod deviceMethod;
 
     private final Map<IotHubTransportMessage, Integer> receivedMessagesToAcknowledge = new ConcurrentHashMap<>();
-    private final Map<Integer, Message> unacknowledgedSentMessages = new ConcurrentHashMap<>();
 
     /**
      * Constructs an instance from the given {@link DeviceClientConfig}
      * object.
      *
      * @param config the client configuration.
      */
-    public MqttIotHubConnection(DeviceClientConfig config) throws IllegalArgumentException
+    // The warning is for how getSasTokenAuthentication() may return null, but the check that our config uses SAS_TOKEN
+    // auth is sufficient at confirming that getSasTokenAuthentication() will return a non-null instance
+    @SuppressWarnings("ConstantConditions")
+    public MqttIotHubConnection(DeviceClientConfig config) throws IllegalArgumentException, TransportException
     {
-        synchronized (MQTT_CONNECTION_LOCK)
+        if (config == null)
+        {
+            throw new IllegalArgumentException("The DeviceClientConfig cannot be null.");
+        }
+        if (config.getIotHubHostname() == null || config.getIotHubHostname().length() == 0)
+        {
+            throw new IllegalArgumentException("hostName cannot be null or empty.");
+        }
+        if (config.getDeviceId() == null || config.getDeviceId().length() == 0)
+        {
+            throw new IllegalArgumentException("deviceID cannot be null or empty.");
+        }
+
+        this.config = config;
+
+        SSLContext sslContext;
+        try
+        {
+            sslContext = this.config.getAuthenticationProvider().getSSLContext();
+        }
+        catch (IOException e)
+        {
+            throw new TransportException("Failed to get SSLContext", e);
+        }
+
+        char[] password = null;
+        if (this.config.getAuthenticationType() == DeviceClientConfig.AuthType.SAS_TOKEN)
```

Comment:
A lot of this constructor logic was moved up from open time to constructor time so that the deviceMessaging/deviceTwin/deviceMethod fields could be declared final to help avoid NPEs

Reply:
No IO operations were moved for this, though. Just initialization logic for the mqtt client that this layer uses